### ISSUE TITLE
Fine-Grained Control Over TOCO Quantization

### ISSUE THREAD

**Issue by deepconvneuralnet** [tensorflow/tensorflow#21009]
Our (Syntiant Corp’s) neural network inference chips use quantized weights and biases in order to minimize storage and energy consumption. The new Tensorflow experimental quantization feature [tf.contrib.quantize.experimental_create_training_graph](https://www.tensorflow.org/api_docs/python/tf/contrib/quantize/experimental_create_training_graph) supports quantizing weights to between 2 and n bits, but the [tf.contrib.lite.toco_convert](https://www.tensorflow.org/api_docs/python/tf/contrib/lite/toco_convert) tool currently only supports 8 bit quantization. As a result, we have to internally fork the TFLite pipeline before generating the Flatbuffer.

Feature request: Update TOCO to support arbitrary (i.e., 2 to n bit) signed fixed point quantization of weights and biases for both symmetric and asymmetric quantization. Our desired solution would process the quantization [specified at the op or Keras layer level](https://github.com/tensorflow/tensorflow/issues/21008) and not involve quantization specification within the TOCO tool API.

**Comment by aselle** [comment]
We are thinking about how to handle things things like this, but a solution is not currently available. It's likely this will be quite difficult. @raghuraman-k , @suharshs , could you please take a look.

**Comment by suharshs** [comment]
This will likely involve needing to introduce new inference types to TOCO/TFLite. It will be a non-trivial change, as we will likely need to also support mixed precision model to make this not be very detrimental to accuracy. This is something we will try to determine a priority for. Thanks!

**Comment by tensorflowbutler** [comment]
Nagging Assignees @suharshs, @raghuraman-k: It has been 17 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @suharshs, @raghuraman-k: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @suharshs, @raghuraman-k: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @suharshs, @raghuraman-k: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @suharshs, @raghuraman-k: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 104 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by suharshs** [comment]
We are working to make things more configurable in new converter iterations, will close this issue since its a bit stale.
